File: 프로그래머스/Lv2/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

vector<int> solution(vector<string> operations) {
    vector<int> answer(2,0);
    int countI=0;
    int countD=0;
    priority_queue <int, vector<int> > maxHeap;
    priority_queue <int, vector<int>, greater<int> > minHeap;
    map <int, int> m;
    
    for(int i=0;i<operations.size();i++){
        if(operations[i][0]=='I'){ //삽입
            countI++;
            int num = stoi(operations[i].substr(2,operations[i].size()));
            maxHeap.push(num);
            minHeap.push(num);
            
        }else if(operations[i][2]=='-'){ //최솟값 삭제
            countD++;
            if(countD==countI){
                while(!minHeap.empty()) minHeap.pop();
                while(!maxHeap.empty()) maxHeap.pop();
                countD=0, countI=0;
            }
            if(!minHeap.empty()) minHeap.pop();
            
        }else{ //최댓값 삭제
            countD++;
            if(countD==countI){
                while(!minHeap.empty()) minHeap.pop();
                while(!maxHeap.empty()) maxHeap.pop();
                countD=0, countI=0;
            }
            if(!maxHeap.empty()) maxHeap.pop();
        }
    }
    
    while(!maxHeap.empty()){
        m[maxHeap.top()]++;
        maxHeap.pop();
    }
    while(!minHeap.empty()){
        m[minHeap.top()]++;
        minHeap.pop();
    }
    
    auto iter = m.begin();
    int max=-99999999;
    int min=99999999;
    while(iter!=m.end()){
        if(iter->second>=2){ //존재하는 값
            int num = iter->first;
            if(num<min) min=num;
            if(num>max) max=num;
        }
        iter++;
    }
    if(min==99999999||max==-99999999){
        return answer;
    }
    if(min==max){
        if(min<0) answer[1]=min;
        else answer[0]=min;
    }
    else{
        answer[0]=max;
        answer[1]=min;
    }
    
    return answer;
}

File: 프로그래머스/Lv3/이중우선순위큐.py
import heapq

def solution(operations):
    answer = []
    d={}
    max_heap=[]
    min_heap=[]
    
    for str in operations:
        command=str[0]
        num=int(str[2:])
        
        if(command=='I'):
            if num in d: d[num]= d[num]+1
            else: d[num]=1
            heapq.heappush(max_heap,num*(-1))
            heapq.heappush(min_heap,num)
        else:
            if(len(max_heap)==0 or len(min_heap)==0):
                max_heap=[]
                min_heap=[]
                continue
            if(num==1):
                while(len(max_heap)>0):
                    q_num=heapq.heappop(max_heap)*(-1)
                    if(q_num in d and d[q_num]<=0): continue
                    if(q_num in d ): d[q_num]-=1
                    break
            else:
               while(len(min_heap)>0):
                    q_num=heapq.heappop(min_heap)
                    if(q_num in d and d[q_num]<=0): continue
                    if(q_num in d ): d[q_num]-=1
                    break 
            
    
    while(len(min_heap)>0):
        q_num=heapq.heappop(min_heap)
        if(q_num in d and d[q_num]<=0): continue
        if(q_num in d): answer.append(q_num)
        break 
        
    while(len(max_heap)>0):
        q_num=heapq.heappop(max_heap)*(-1)
        if(q_num in d and d[q_num]<=0): continue
        if(q_num in d): answer.append(q_num)
        break 
    
    if(len(answer)==0): answer=[0,0]
    return sorted(answer, reverse=True)
